runtime/handler.go

```diff
@@ -168,48 +168,41 @@ func handleForwardResponseOptions(ctx context.Context, w http.ResponseWriter, re
 	return nil
 }
 
-func handleForwardResponseStreamError(wroteHeader bool, marshaler Marshaler, w http.ResponseWriter, err error) {
-	buf, merr := marshaler.Marshal(streamChunk(nil, err))
+func handleForwardResponseStreamError(ctx context.Context, wroteHeader bool, marshaler Marshaler, w http.ResponseWriter, req *http.Request, mux *ServeMux, err error) {
+	if !wroteHeader {
+		// if we haven't already written header, use normal error handler to render
+		HTTPError(ctx, mux, marshaler, w, req, err)
+		return
```

Comment:
Is this backwards compatibility breaking? I can see we used to do just `w.WriteHeader` here.

Reply:
I guess technically it is (though the behavior in this case isn't really specified, at least not that I can find).

Previously, it would emit a JSON response body like so:
```json
{
  "error": {
    "grpc_code": 13,
    "http_code": 500,
    "message": "some error occurred",
    "http_status": "Internal Server Error",
    "details": [],
  }
}
```

Now, it will emit a response body like this instead:

```json
{
  "code": 13,
  "message": "some error occurred",
  "details": [],
}
```

The actual HTTP status code sent back would be the same (in the example: "500 Internal Server Error").

IMO, this makes it cohesive and consistent: if the client gets a non-200 status code, the response format is the same (regardless of whether the method was a streaming or unary method). So it feels more like a fix and will also simplify the specification/documentation of error handling.

But if that is too great of a change, how about the following alternative logic:

1. Call `streamErrorChunk` to let custom handler produce custom stream error message.
2. Use the resulting error's `HttpCode` to set the response code (if `!wroteHeader`).

Reply:
The swagger generator defines the error format for streams including the envelope, so I would prefer the consistency, even though I understand your point about the non-200 error code response.

Reply:
> The swagger generator defines the error format for streams including the envelope

Oh, it does? I guess that must be another fix that isn't in our old fork. (When I run the swagger protoc plugin in our repo, it emits the wrong definition.)

Okay, in that case, I'll tweak this behavior.

Reply:
> The swagger generator defines the error format for streams including the envelope

Out of curiosity, is this spec meant for interpreting responses for _successful_ calls only? Does swagger even allow specifying different response types based on error-vs-ok conditions? Because currently, unary RPCs may not be correctly encoded in swagger for non-200 response codes. (Maybe that's only in my old fork of the repo?)

Reply:
It's an optional field in the response, so the swagger definition shows the "result" key and the "error" for all responses.